src/Microsoft.VisualStudio.Services.Agent/ProcessInvoker.cs

```diff
@@ -194,41 +177,132 @@ private void Dispose(bool disposing)
-                _processExitedSignal.Dispose();
                 if (_proc != null)
                 {
-                    _proc.Exited -= OnExited;
-                    _proc.ErrorDataReceived -= OnErrorDataReceived;
-                    _proc.OutputDataReceived -= OnOutputDataReceived;
                     _proc.Dispose();
                     _proc = null;
                 }
             }
         }
 
-        private void OnExited(object sender, EventArgs e)
+        private void ProcessOutput()
         {
-            _stopWatch.Stop();
-            _processExitedSignal.Release();
+            List<string> errorData = new List<string>();
+            List<string> outputData = new List<string>();
+
+            string errorLine;
+            while (_errorData.TryDequeue(out errorLine))
+            {
+                errorData.Add(errorLine);
+            }
+
+            string outputLine;
+            while (_outputData.TryDequeue(out outputLine))
+            {
+                outputData.Add(outputLine);
+            }
+
+            _outputProcessEvent.Reset();
+
+            // Write the error lines.
+            if (errorData != null)
+            {
+                foreach (string line in errorData)
+                {
+                    if (line != null && this.ErrorDataReceived != null)
+                    {
+                        this.ErrorDataReceived(this, new ProcessDataReceivedEventArgs(line));
+                    }
+                }
+            }
+
+            // Process the output lines.
+            if (outputData != null)
+            {
+                foreach (string line in outputData)
+                {
+                    if (line != null && this.OutputDataReceived != null)
+                    {
+                        // The line is output from the process that was invoked.
+                        this.OutputDataReceived(this, new ProcessDataReceivedEventArgs(line));
+                    }
+                }
+            }
         }
 
-        private void OnErrorDataReceived(object sender, DataReceivedEventArgs e)
+        private void CancelProcessTree()
         {
-            // at the end of the process, the event fires one last time with null
-            if (e.Data != null)
+            ArgUtil.NotNull(_proc, nameof(_proc));
+
+            try
+            {
+                // TODO: Send Ctrl+C/Break to process group.
+                if (!_proc.HasExited)
+                {
+                    _proc.Kill();
+                }
+            }
+            catch (InvalidOperationException)
             {
-                ErrorDataReceived?.Invoke(sender, e);
+                // InvalidOperationException can occur if process got terminated by itself between 
+                // HasExited and Kill() calls above.
             }
         }
 
-        private void OnOutputDataReceived(object sender, DataReceivedEventArgs e)
+        private void ProcessExitedHandler(object sender, EventArgs e)
```

Comment:
Can you add a comment here what this solving? I don't understand why data is still being written to the streams after the process exits. Is the exit event misleading (e.g. fired before the process really exits)? Are child processes still writing to the streams (if so, is waiting for the process tree to exit an option instead)? Or is this purely because STDOUT/STDERR are being pumped on separate threads?

It looks like the _asyncStreamReaderCount may have a chance to drop to zero before 5 seconds is up. Rather than add 5 seconds to every process invocation, would it be better wait in a while loop up to a total of X seconds, and delay for only a small amount each iteration?